pkg/sfu/downtrack.go

```diff
@@ -502,56 +502,52 @@ func (d *DownTrack) Close() {
 	})
 }
 
-func (d *DownTrack) CurrentSpatialLayer() int32 {
-	return d.currentSpatialLayer.get()
-}
-
-func (d *DownTrack) TargetSpatialLayer() int32 {
-	return d.targetSpatialLayer.get()
-}
-
-func (d *DownTrack) SetMaxSpatialLayer(spatialLayer int32) error {
-	// LK-TODO: support SVC
-	if d.trackType != SimulcastDownTrack {
-		return ErrSpatialNotSupported
-	}
-
-	if spatialLayer == d.MaxSpatialLayer() {
-		return nil
+func (d *DownTrack) SetMaxSpatialLayer(spatialLayer int32) {
+	d.translationMu.Lock()
+	if spatialLayer == d.maxSpatialLayer {
+		d.translationMu.Unlock()
+		return
 	}
 
-	d.maxSpatialLayer.set(spatialLayer)
+	d.maxSpatialLayer = spatialLayer
+	muted := d.muted
+	maxTemporalLayer := d.maxTemporalLayer
+	d.translationMu.Unlock()
 
-	if d.enabled.get() && d.onSubscribedLayersChanged != nil {
-		d.onSubscribedLayersChanged(d, spatialLayer, d.MaxTemporalLayer())
+	if !muted && d.onSubscribedLayersChanged != nil {
+		d.onSubscribedLayersChanged(d, spatialLayer, maxTemporalLayer)
 	}
-
-	return nil
-}
-
-func (d *DownTrack) MaxSpatialLayer() int32 {
-	return d.maxSpatialLayer.get()
 }
 
 func (d *DownTrack) SetMaxTemporalLayer(temporalLayer int32) {
-	if temporalLayer == d.MaxTemporalLayer() {
+	d.translationMu.Lock()
+	if temporalLayer == d.maxTemporalLayer {
+		d.translationMu.Unlock()
 		return
 	}
 
-	d.maxTemporalLayer.set(temporalLayer)
+	d.maxTemporalLayer = temporalLayer
+	muted := d.muted
+	maxSpatialLayer := d.maxSpatialLayer
+	d.translationMu.Unlock()
 
-	if d.onSubscribedLayersChanged != nil {
-		d.onSubscribedLayersChanged(d, d.MaxSpatialLayer(), temporalLayer)
+	if !muted && d.onSubscribedLayersChanged != nil {
+		d.onSubscribedLayersChanged(d, maxSpatialLayer, temporalLayer)
 	}
 }
 
-func (d *DownTrack) MaxTemporalLayer() int32 {
-	return d.maxTemporalLayer.get()
+func (d *DownTrack) MaxLayers() (int32, int32) {
+	d.translationMu.RLock()
+	defer d.translationMu.RUnlock()
+
+	return d.maxSpatialLayer, d.maxTemporalLayer
 }
 
-// switchSpatialLayer switches the target layer
-func (d *DownTrack) switchSpatialLayer(targetLayer int32) {
-	d.targetSpatialLayer.set(targetLayer)
+func (d *DownTrack) IsForwardingOptimal() bool {
```

Comment:
What about returning a enum for ForwardingStatus? `Optimal, Partial, Off` ? that'd map cleanly to the three  quality levels.

Reply:
Cool, can do that.

Reply:
@davidzhao Addressed in this commit - https://github.com/livekit/livekit-server/pull/213/commits/cc2d11f41ab24f972b866d7f7b22de29ea3259b8. Also, addressed Jie's comments and split out a `Forwarder` structure to keep things cleaner and easier to follow.